fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/InventoryWrappersImpl.java

```diff
@@ -0,0 +1,458 @@
+						.mapToObj(slot -> new SidedSlotWrapper(slots.get(slot), sidedInventory, direction))
+						.collect(Collectors.toList());
+				result.add(new CombinedStorage<>(sideSlots));
+			}
+		} else {
+			// unsided logic, just use the same Storage 7 times
+			for (int i = 0; i < 6; ++i) { // 6 directions
+				result.add(fullWrapper);
+			}
+		}
+
+		result.add(fullWrapper);
+		return result;
+	}
+
+	// Wraps a single slot of an unsided Inventory.
+	private static class SlotWrapper implements Storage<ItemKey>, StorageView<ItemKey>, Participant<ItemStack> {
+		private final Inventory inventory;
+		private final int slot;
+
+		private SlotWrapper(Inventory inventory, int slot) {
+			this.inventory = inventory;
+			this.slot = slot;
+		}
+
+		@Override
+		public boolean supportsInsertion() {
+			return true;
+		}
+
+		@Override
+		public long insert(ItemKey key, long maxAmount, Transaction transaction) {
+			// TODO: clean this up
+			ItemPreconditions.notEmpty(key);
+			int count = (int) Math.min(maxAmount, inventory.getMaxCountPerStack());
+			ItemStack stack = inventory.getStack(slot);
+
+			if (stack.isEmpty()) {
+				ItemStack keyStack = key.toStack(count);
+
+				if (inventory.isValid(slot, keyStack)) {
+					int inserted = Math.min(keyStack.getMaxCount(), count);
+					transaction.enlist(this);
+					keyStack.setCount(inserted);
+					inventory.setStack(slot, keyStack);
+					return inserted;
+				}
+			} else if (key.matches(stack)) {
+				int inserted = Math.min(stack.getMaxCount() - stack.getCount(), count);
+				transaction.enlist(this);
+				stack.increment(inserted);
+				return inserted;
+			}
+
+			return 0;
+		}
+
+		@Override
+		public boolean supportsExtraction() {
+			return true;
+		}
+
+		@Override
+		public long extract(ItemKey key, long maxAmount, Transaction transaction) {
+			ItemPreconditions.notEmpty(key);
+			ItemStack stack = inventory.getStack(slot);
+
+			if (key.matches(stack)) {
+				int extracted = (int) Math.min(stack.getCount(), maxAmount);
+				transaction.enlist(this);
+				stack.decrement(extracted);
+				return extracted;
+			}
+
+			return 0;
+		}
+
+		@Override
+		public boolean forEach(Visitor<ItemKey> visitor, Transaction transaction) {
+			if (!inventory.getStack(slot).isEmpty()) {
+				return visitor.accept(this);
+			}
+
+			return false;
+		}
+
+		@Override
+		public ItemKey resource() {
+			return ItemKey.of(inventory.getStack(slot));
+		}
+
+		@Override
+		public long amount() {
+			return inventory.getStack(slot).getCount();
+		}
+
+		@Override
+		public ItemStack onEnlist() {
+			return inventory.getStack(slot).copy();
+		}
+
+		@Override
+		public void onClose(ItemStack state, TransactionResult result) {
+			if (result.wasAborted()) {
+				inventory.setStack(slot, state);
+			}
+		}
+
+		@Override
+		public void onFinalCommit() {
+			// TODO: is this necessary for player inventories? in what circumstances?
+			/*if (inventory instanceof PlayerInventory) {
```

Comment:
Do any of the set stack methods in player inventory do any packet logic? That would probably determine this answer.

Reply:
I think `offerOrDrop` does it. But frankly I am not satisfied of `PlayerInventoryWrapper` and friends, that still needs to be discussed.

Reply:
I don't think it's necessary in the end, the need stems from a flawed Mojang implementation as far as I can tell.